diff -up nautilus-2.28.4/libnautilus-private/nautilus-desktop-directory.c.bak nautilus-2.28.4/libnautilus-private/nautilus-desktop-directory.c
--- nautilus-2.28.4/libnautilus-private/nautilus-desktop-directory.c.bak	2012-03-01 20:19:49.000000000 +0100
+++ nautilus-2.28.4/libnautilus-private/nautilus-desktop-directory.c	2012-03-01 20:20:57.000000000 +0100
@@ -255,6 +255,15 @@ desktop_cancel_callback (NautilusDirecto
 			(node->data,
 			 directory_ready_callback, merged_callback);
 	}
+	
+	/* Cancel the backing directory */
+	nautilus_directory_cancel_callback_internal
+		(directory,
+		 NULL,
+		 directory_ready_callback,
+		 NULL,
+		 merged_callback);
+	
 	merged_callback_destroy (merged_callback);
 }
 
@@ -459,6 +468,16 @@ forward_files_changed_cover (NautilusDir
 	nautilus_directory_emit_files_changed (NAUTILUS_DIRECTORY (callback_data), files);
 }
 
+static void
+cancel_scheduled_ready_callbacks (gpointer key,
+                                  MergedCallback *merged_callback,
+                                  NautilusDesktopDirectory *desktop)
+{
+  desktop_cancel_callback (NAUTILUS_DIRECTORY (desktop),
+                           merged_callback->callback,
+                           merged_callback->callback_data);
+}
+
 void
 nautilus_desktop_directory_update (NautilusDesktopDirectory *desktop)
 {
@@ -467,7 +486,11 @@ nautilus_desktop_directory_update (Nauti
 	NautilusDirectory *real_directory;
 
 	real_directory = desktop->details->real_directory;
+	
 	if (real_directory != NULL) {
+		/* Cancels all pending events */
+		g_hash_table_foreach (desktop->details->callbacks, (GHFunc) cancel_scheduled_ready_callbacks, desktop);
+	
 		g_hash_table_foreach_remove (desktop->details->callbacks, (GHRFunc) gtk_true, NULL);
 		g_hash_table_foreach_remove (desktop->details->monitors, (GHRFunc) gtk_true, NULL);
 
