diff -up nautilus-2.28.4/libnautilus-private/nautilus-file-operations.c.bak nautilus-2.28.4/libnautilus-private/nautilus-file-operations.c
--- nautilus-2.28.4/libnautilus-private/nautilus-file-operations.c.bak	2011-01-13 16:23:46.000000000 +0100
+++ nautilus-2.28.4/libnautilus-private/nautilus-file-operations.c	2011-01-13 16:25:14.000000000 +0100
@@ -802,7 +802,7 @@ custom_basename_to_string (char *format,
 	info = g_file_query_info (file,
 				  G_FILE_ATTRIBUTE_STANDARD_DISPLAY_NAME,
 				  0,
-				  g_cancellable_get_current (),
+				  NULL,
 				  NULL);
 	
 	name = NULL;
