diff -up nautilus-2.28.4/libnautilus-private/nautilus-file.c.bak nautilus-2.28.4/libnautilus-private/nautilus-file.c
--- nautilus-2.28.4/libnautilus-private/nautilus-file.c.bak	2012-03-02 15:22:27.000000000 +0100
+++ nautilus-2.28.4/libnautilus-private/nautilus-file.c	2012-03-02 15:23:42.000000000 +0100
@@ -6713,7 +6713,6 @@ get_fs_free_cb (GObject *source_object,
 			nautilus_file_unref (file);
 		}
 	}
-	nautilus_directory_unref (directory);
 }
 
 /**
@@ -6731,7 +6730,7 @@ nautilus_file_get_volume_free_space (Nau
 	char *res;
 	time_t now;
 
-	directory = nautilus_directory_get_for_file (file);
+	directory = file->details->directory;
 
 	now = time (NULL);
 	/* Update first time and then every 2 seconds */
@@ -6745,8 +6744,6 @@ nautilus_file_get_volume_free_space (Nau
 						    get_fs_free_cb,
 						    directory); /* Inherits ref */
 		g_object_unref (location);
-	} else {
-		nautilus_directory_unref (directory);
 	}
 
 
