diff -up nautilus-2.28.4/libnautilus-private/nautilus-bookmark.c.bak nautilus-2.28.4/libnautilus-private/nautilus-bookmark.c
--- nautilus-2.28.4/libnautilus-private/nautilus-bookmark.c.bak	2009-04-20 12:57:19.000000000 +0200
+++ nautilus-2.28.4/libnautilus-private/nautilus-bookmark.c	2011-01-18 15:32:09.000000000 +0100
@@ -303,7 +303,7 @@ nautilus_bookmark_set_name (NautilusBook
 	g_free (bookmark->details->name);
 	bookmark->details->name = g_strdup (new_name);
 
-	g_signal_emit (bookmark, signals[APPEARANCE_CHANGED], 0);
+	g_signal_emit (bookmark, signals[CONTENTS_CHANGED], 0);
 
 	return TRUE;
 }
