--- b/libnautilus-private/nautilus-file.c	2014-03-04 15:13:36.661778869 +0530
+++ a/libnautilus-private/nautilus-file.c	2014-03-04 15:15:31.820452414 +0530
@@ -541,7 +541,7 @@
 	GList **list_ptr;
 
 	/* Check if there is a symlink name. If none, we are OK. */
-	if (file->details->symlink_name == NULL) {
+	if (file->details->symlink_name == NULL || !nautilus_file_is_symbolic_link (file)) {
 		return;
 	}
 
