diff -up nautilus-2.28.4/libnautilus-private/nautilus-desktop-directory.c.xdg-refresh nautilus-2.28.4/libnautilus-private/nautilus-desktop-directory.c
--- nautilus-2.28.4/libnautilus-private/nautilus-desktop-directory.c.xdg-refresh	2009-12-08 16:17:07.000000000 +0100
+++ nautilus-2.28.4/libnautilus-private/nautilus-desktop-directory.c	2011-01-19 16:32:14.000000000 +0100
@@ -32,6 +32,7 @@
 #include "nautilus-file-private.h"
 #include "nautilus-file-utilities.h"
 #include "nautilus-global-preferences.h"
+#include "nautilus-signaller.h"
 #include <eel/eel-glib-extensions.h>
 #include <eel/eel-gtk-macros.h>
 #include <gtk/gtk.h>
@@ -458,8 +459,8 @@ forward_files_changed_cover (NautilusDir
 	nautilus_directory_emit_files_changed (NAUTILUS_DIRECTORY (callback_data), files);
 }
 
-static void
-update_desktop_directory (NautilusDesktopDirectory *desktop)
+void
+nautilus_desktop_directory_update (NautilusDesktopDirectory *desktop)
 {
 	char *desktop_path;
 	char *desktop_uri;
@@ -496,7 +497,7 @@ update_desktop_directory (NautilusDeskto
 static void
 desktop_directory_changed_callback (gpointer data)
 {
-	update_desktop_directory (NAUTILUS_DESKTOP_DIRECTORY (data));
+	nautilus_desktop_directory_update (NAUTILUS_DESKTOP_DIRECTORY (data));
 	nautilus_directory_force_reload (NAUTILUS_DIRECTORY (data));
 }
 
@@ -511,7 +512,7 @@ nautilus_desktop_directory_init (Nautilu
 	desktop->details->monitors = g_hash_table_new_full (NULL, NULL,
 							    NULL, (GDestroyNotify)merged_monitor_destroy);
 
-	update_desktop_directory (NAUTILUS_DESKTOP_DIRECTORY (desktop));
+	nautilus_desktop_directory_update (NAUTILUS_DESKTOP_DIRECTORY (desktop));
 
 	eel_preferences_add_callback (NAUTILUS_PREFERENCES_DESKTOP_IS_HOME_DIR,
 				      desktop_directory_changed_callback,
diff -up nautilus-2.28.4/libnautilus-private/nautilus-desktop-directory.h.xdg-refresh nautilus-2.28.4/libnautilus-private/nautilus-desktop-directory.h
--- nautilus-2.28.4/libnautilus-private/nautilus-desktop-directory.h.xdg-refresh	2009-12-08 10:50:42.000000000 +0100
+++ nautilus-2.28.4/libnautilus-private/nautilus-desktop-directory.h	2011-01-19 16:32:14.000000000 +0100
@@ -55,5 +55,7 @@ typedef struct {
 
 GType   nautilus_desktop_directory_get_type             (void);
 NautilusDirectory * nautilus_desktop_directory_get_real_directory   (NautilusDesktopDirectory *desktop_directory);
+void    nautilus_desktop_directory_update               (NautilusDesktopDirectory *desktop);
+
 
 #endif /* NAUTILUS_DESKTOP_DIRECTORY_H */
diff -up nautilus-2.28.4/src/nautilus-desktop-window.c.xdg-refresh nautilus-2.28.4/src/nautilus-desktop-window.c
--- nautilus-2.28.4/src/nautilus-desktop-window.c.xdg-refresh	2009-12-11 13:31:50.000000000 +0100
+++ nautilus-2.28.4/src/nautilus-desktop-window.c	2011-01-19 16:34:15.000000000 +0100
@@ -34,6 +34,8 @@
 #include <eel/eel-vfs-extensions.h>
 #include <libnautilus-private/nautilus-file-utilities.h>
 #include <libnautilus-private/nautilus-icon-names.h>
+#include <libnautilus-private/nautilus-signaller.h>
+#include <libnautilus-private/nautilus-desktop-directory.h>
 #include <gio/gio.h>
 #include <glib/gi18n.h>
 
@@ -84,9 +86,15 @@ void
 nautilus_desktop_window_update_directory (NautilusDesktopWindow *window)
 {
 	GFile *location;
+	NautilusDirectory *directory;
 	
 	g_assert (NAUTILUS_IS_DESKTOP_WINDOW (window));
 	
+	directory = nautilus_directory_get_by_uri (EEL_DESKTOP_URI);
+	if (directory && NAUTILUS_IS_DESKTOP_DIRECTORY (directory)) {
+		nautilus_desktop_directory_update (NAUTILUS_DESKTOP_DIRECTORY (directory));
+	}
+	
 	NAUTILUS_SPATIAL_WINDOW (window)->affect_spatial_window_on_next_location_change = TRUE;
 	location = g_file_new_for_uri (EEL_DESKTOP_URI);
 	nautilus_window_go_to (NAUTILUS_WINDOW (window), location);
@@ -136,6 +144,11 @@ nautilus_desktop_window_new (NautilusApp
 	 */
 	nautilus_desktop_window_update_directory (window);
 
+	/* Refresh desktop on xdg dirs change */
+	g_signal_connect_object (nautilus_signaller_get_current (),
+				 "user_dirs_changed",
+				 G_CALLBACK (nautilus_desktop_window_update_directory),
+				 window, G_CONNECT_SWAPPED);
 	return window;
 }
 
diff -up nautilus-2.28.4/src/nautilus-window-manage-views.c.xdg-refresh nautilus-2.28.4/src/nautilus-window-manage-views.c
--- nautilus-2.28.4/src/nautilus-window-manage-views.c.xdg-refresh	2009-12-15 10:10:43.000000000 +0100
+++ nautilus-2.28.4/src/nautilus-window-manage-views.c	2011-01-19 16:46:21.000000000 +0100
@@ -48,6 +48,7 @@
 #include <eel/eel-gtk-macros.h>
 #include <eel/eel-stock-dialogs.h>
 #include <eel/eel-string.h>
+#include <eel/eel-vfs-extensions.h>
 #include <gtk/gtk.h>
 #include <gdk/gdkx.h>
 #include <glib/gi18n.h>
@@ -486,6 +487,7 @@ nautilus_window_slot_open_location_full 
         NautilusWindowSlot *target_slot;
 	NautilusWindowOpenFlags slot_flags;
         gboolean do_load_location = TRUE;
+        gboolean is_desktop;
 	GFile *old_location;
 	char *old_uri, *new_uri;
 	int new_slot_position;
@@ -505,6 +507,7 @@ nautilus_window_slot_open_location_full 
 			    window,
 			    old_uri,
 			    new_uri);
+	is_desktop = eel_uri_is_desktop (new_uri);
 	g_free (old_uri);
 	g_free (new_uri);
 
@@ -615,7 +618,8 @@ nautilus_window_slot_open_location_full 
 
         if ((!do_load_location) ||
 	    (target_window == window && target_slot == slot &&
-	     old_location && g_file_equal (old_location, location))) {
+	     old_location && g_file_equal (old_location, location) &&
+	     ! is_desktop)) {
 		if (old_location) {
 			g_object_unref (old_location);
 		}
